collections/src/main/scala/strawman/collection/LinearSeq.scala

```diff
@@ -171,4 +171,8 @@ trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A]] extends Any w
     }
     last
   }
+
+  override def tails: Iterator[C] =
+    Iterator.iterate(coll)(_.tail.asInstanceOf[C]).takeWhile(_.nonEmpty) ++ Iterator(newSpecificBuilder().result())
```

Comment:
Is the cast necessary?

Reply:
Yes, the static return type of `tail` is `LinearSeq[A]`, whereas we want a `C`.

Reply:
Hm, that's not nice. We lose the precise type on repeated applications of `tail`. `this.type.tail` results in C but `C.type` is only `LinearSeq[A]`.

Reply:
Having a more precise type would require a more complex upper bound on `C`.

Reply:
The problem lies in the definition of `LinearSeqOps`. We need to make `C` recursive:
```scala
trait LinearSeqOps[+A, +CC[X] <: LinearSeq[X], +C <: LinearSeq[A] with LinearSeqOps[A, CC, C]] extends Any with SeqOps[A, CC, C]
```
And the same for `immutable.LinearSeqOps` and `LazyListOps`. With these changes the cast is no longer needed.

Reply:
Yes, that’s what I was trying to say in my previous comment. The drawback is that it becomes harder to abstract over _some_ `LinearSeqOps`. See how it is done for `immutable.MapOps`, which has similar bounds: https://github.com/scala/collection-strawman/blob/master/collections-contrib/src/main/scala/strawman/collection/decorators/HasImmutableMapOps.scala#L12-L17

Reply:
I don't think this is a problem for `LinearSeqOps`. The only changes I had to do downstream was to repeat the bounds in `immutable.LinearSeqOps` and `LazyListOps`. `LinearSeqOps` is just an implementation detail. You'd abstract over `SeqOps`, possibly restricted to subtypes of `LinearSeq` if you care about performance characteristics.

Reply:
Allright, I’ve fixed that in a304aaf.